internal/checker/printer.go

```diff
@@ -348,7 +348,27 @@ func (p *Printer) printTupleType(t *Type) {
 	p.print("]")
 }
 
+func (c *Checker) isClassInstanceSide(t *Type) bool {
+	return t.symbol != nil && t.symbol.Flags&ast.SymbolFlagsClass != 0 && (t == c.getDeclaredTypeOfClassOrInterface(t.symbol) || (t.flags&TypeFlagsObject != 0 && t.objectFlags&ObjectFlagsIsClassInstanceClone != 0))
+}
+
 func (p *Printer) printAnonymousType(t *Type) {
+	// !!! logic partially taken from createAnonymousTypeNode
```

Comment:
I think this all can be simplified to the following (which also handles `typeof import`):

```Go
	if t.symbol != nil && len(t.symbol.Name) != 0 {
		if t.symbol.Flags&(ast.SymbolFlagsClass|ast.SymbolFlagsEnum|ast.SymbolFlagsValueModule) != 0 {
			if t == p.c.getTypeOfSymbol(t.symbol) {
				p.print("typeof ")
				if t.symbol.Flags&ast.SymbolFlagsValueModule != 0 && t.symbol.Name[0] == '"' {
					p.print("import(")
					p.print(t.symbol.Name)
					p.print(")")
				} else {
					p.printName(t.symbol)
				}
				return
			}
		}
	}
```

Reply:
I tried it locally and the changes look pretty good, but not sure if they square with yours.

Reply:
Yeah, that seems to work better. Will do that.

Reply:
Wow, that brought the PR from 2.5k lines removed to 5k lines removed. Very nice.